File: ruby/lib/runapi/recraft/resources/upscales.rb
# frozen_string_literal: true

module RunApi
  module Recraft
    module Resources
      class Upscales
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/recraft/upscales"

        RESPONSE_CLASS = Types::ImageTaskResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedImageTaskResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          raise Core::ValidationError, "model is required" unless param(params, :model)
          raise Core::ValidationError, "image_url is required" unless param(params, :image_url)

          model = param(params, :model)
          unless Types::UPSCALE_MODELS.include?(model)
            raise Core::ValidationError, "Invalid model: #{model}. Must be one of: #{Types::UPSCALE_MODELS.join(", ")}"
          end
        end
      end
    end
  end
end

File: js/src/resources/upscales.ts
import type { HttpClient, PollingOptions, RequestOptions } from '@runapi.ai/core';
import { compactParams, ValidationError } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import type { CompletedImageTaskResponse, ImageTaskResponse, TaskCreateResponse, UpscaleParams } from '../types';

const ENDPOINT = '/api/v1/recraft/upscales';

export class Upscales {
  constructor(private readonly http: HttpClient) {}

  async run(params: UpscaleParams, options?: RequestOptions & PollingOptions): Promise<CompletedImageTaskResponse> {
    const { id } = await this.create(params, options);
    const response = await pollUntilComplete<ImageTaskResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
    return response as CompletedImageTaskResponse;
  }

  async create(params: UpscaleParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    const body = compactParams(params);
    if (!body.model) throw new ValidationError('model is required');
    if (!body.image_url) throw new ValidationError('image_url is required');

    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body,
      ...options,
    });
  }

  async get(id: string, options?: RequestOptions): Promise<ImageTaskResponse> {
    return this.http.request<ImageTaskResponse>('GET', `${ENDPOINT}/${id}`, {
      ...options,
    });
  }
}
